### PR TITLE
[SPARK-27521][SQL] Move data source v2 to catalyst module

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently we are in a strange status that, some data source v2 interfaces(catalog related) are in sql/catalyst, some data source v2 interfaces(Table, ScanBuilder, DataReader, etc.) are in sql/core.

I don't see a reason to keep data source v2 API in 2 modules. If we should pick one module, I think sql/catalyst is the one to go.

Catalyst module already has some user-facing stuff like DataType, Row, etc. And we have to update `Analyzer` and `SessionCatalog` to support the new catalog plugin, which needs to be in the catalyst module.

This PR can solve the problem we have in https://github.com/apache/spark/pull/24246

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
No matter how we are going to change the package organization, I think there is no objection to move data source v2 to sql/catalyst.

cc @rxin @rdblue @mccheah @gatorsmile @gengliangwang

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan . Could you fix the build error?
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/src/main/java/org/apache/spark/sql/sources/v2/TableProvider.java:21:  error: cannot find symbol
[error] import org.apache.spark.sql.sources.DataSourceRegister;
[error]           
```

**Comment by HeartSaVioR** [comment]
I remind sql/catalyst module is (implicitly) tend to be treated as non-public so leaving classes as public was OK - if we decide to place public APIs to sql/catalyst and treat sql/catalyst module be public, there might be some spots we may want to explicitly hide.

**Comment by mccheah** [comment]
@HeartSaVioR this has been discussed before - I think while that was previously the case, the current push to Data Source V2 necessitates us to change that opinion. There might be a different Maven module organization that is sensible, for example having an `sql-api` or `catalyst-api` module.

**Comment by cloud-fan** [comment]
> I remind sql/catalyst module is (implicitly) tend to be treated as non-public

Actually this is not true, at least the public `DataType` API is in sql/catalyst.

**Review by HyukjinKwon — COMMENTED** [review]
@cloud-fan, can you double check if API doc is appropriately generated, and clarify this PR doesn't target to move under catalyst package, `org/apache/spark/sql/catalyst` in PR description?

**Comment by rdblue** [comment]
Looks like this moved the v2 package and the vectorized package. Is it necessary to move `ArrowColumnVector`? That seems like an implementation class and not an interface.

Also, does it make sense to move `DataSourceRegister`? That seems like a part of v1 that we've just reused. Maybe we should add a method to TableProvider instead of moving the trait?

**Comment by gengliangwang** [comment]
> Looks like this moved the v2 package and the vectorized package. Is it necessary to move ArrowColumnVector? That seems like an implementation class and not an interface.

I think @cloud-fan move the vectorized package because `org.apache.spark.sql.vectorized.ColumnarBatch` is used in `PartitionReaderFactory `.

**Comment by rdblue** [comment]
@gengliangwang, I agree. It looks like the entire package was moved. Is it necessary to move the entire package, or should this move just the interfaces?

**Comment by gengliangwang** [comment]
If it is in the `sql` package, we can't use it catalyst package unless we create new abstraction in API level.  Also, `ColumnarBatch` is a final class.  Though ugly, it seems necessary to me.

**Comment by rdblue** [comment]
Sounds fine to me.

**Comment by dongjoon-hyun** [comment]
Could you resolve the conflicts please? @cloud-fan

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/java/org/apache/spark/sql/execution/ArrowColumnVector.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.spark.sql.vectorized;
+package org.apache.spark.sql.execution;
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/execution/ArrowColumnVector.java]: Can we put `ArrowColumnVector` into `org.apache.spark.sql.vectorized` like `ColumnVector`?

Comment (dongjoon-hyun) [reply sql/core/src/main/java/org/apache/spark/sql/execution/ArrowColumnVector.java]: Most `import` changes seems to be due to this relocation from `vectorized` to `execution`.

File: sql/catalyst/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsDynamicOverwrite.java [hunk 2]
```diff
@@ -25,7 +25,7 @@
  * write does not contain data will remain unchanged.
  * <p>
  * This is provided to implement SQL compatible with Hive table operations but is not recommended.
- * Instead, use the {@link SupportsOverwrite overwrite by filter API} to explicitly replace data.
+ * Instead, use the {@code SupportsOverwrite overwrite by filter API} to explicitly replace data.
```
Comment (rdblue) [inline sql/catalyst/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsDynamicOverwrite.java]: `SupportsOverwrite` should still be available, right?

File: sql/core/src/main/java/org/apache/spark/sql/execution/ArrowColumnVector.java (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/java/org/apache/spark/sql/execution/ArrowColumnVector.java]: I think this depends on whether we want to also reorganize in this commit, or just move from SQL to Catalyst. @mccheah suggested reorganizing at the same time. I don't have a strong opinion on whether we do it at the same time or separately, but if we aren't reorganizing the v2 classes, then I agree that this should not move packages.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/SupportsPushDownFilters.java [hunk 3]
```diff
@@ -25,6 +25,7 @@
  * push down filters to the data source and reduce the size of the data to be read.
  */
 @Evolving
+// TODO: create data source v2 filter API and move this interface to the catalyst module.
```
Comment (rdblue) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/SupportsPushDownFilters.java]: Why can't we move the filter API as well?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala [hunk 4]
```diff
@@ -305,7 +305,7 @@ class RateStreamProviderSuite extends StreamTest {
         .load()
     }
     assert(exception.getMessage.contains(
-      "rate source does not support user-specified schema"))
+      "RateStreamProvider source does not support user-specified schema"))
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala]: This is an unrelated change and should be done in a separate commit.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala [hunk 5]
```diff
@@ -204,7 +204,7 @@ class TextSocketStreamSuite extends StreamTest with SharedSQLContext with Before
       provider.getTable(new CaseInsensitiveStringMap(params.asJava), userSpecifiedSchema)
     }
     assert(exception.getMessage.contains(
-      "socket source does not support user-specified schema"))
+      "TextSocketSourceProvider source does not support user-specified schema"))
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/TextSocketStreamSuite.scala]: Same here. This doesn't need to be done and causes this PR to touch additional files.

File: sql/catalyst/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsDynamicOverwrite.java (hunk 2 above)
Comment (rdblue) [reply sql/catalyst/src/main/java/org/apache/spark/sql/sources/v2/writer/SupportsDynamicOverwrite.java]: Nevermind, I see that it was not moved because the filter API is in the SQL module. We should make a decision about whether we move the public filter API as well. I would prefer to move it to Catalyst.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/sources/RateStreamProviderSuite.scala]: This is because I hesitate to move `DataSourceRegister` to catalyst, and then DS v2 is not able to get `DataSourceRegister#shorName`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala [hunk 6]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.spark.sql.execution.arrow
+package org.apache.spark.sql.util
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/ArrowUtils.scala]: I don't want to add an `execution` package in catalyst, so I changed the pacakge for this internal class during the move.

File: sql/catalyst/pom.xml [hunk 7]
```diff
@@ -114,6 +114,10 @@
       <version>2.7.3</version>
       <type>jar</type>
     </dependency>
+    <dependency>
+      <groupId>org.apache.arrow</groupId>
+      <artifactId>arrow-vector</artifactId>
```
Comment (cloud-fan) [inline sql/catalyst/pom.xml]: Since the `ArrowColumnVector` is moved to the catalyst module, I have to move the dependency as well. I think it's fine as sql/core depends on sql/catalyst.

**Comment by cloud-fan** [comment]
I moved a few public classes from sql/core to sql/catalyst, without changing the package name. I think this is binary compatible as sql/core depends on sql/catalyst.

I'm sure if the mima failures are legitimate,  do you have any ideas? @srowen @JoshRosen

**Comment by srowen** [comment]
The MiMa warning is legit because those classes are no longer in core. catalyst depends on core, right, not the other way around? so anyone that only depends on core won't see these classes anymore. That could be fine as a breaking change for Spark 3, but it's a legit warning.

**Comment by rdblue** [comment]
sql/core depends on sql/catalyst, so the classes should always be there. When looking at sql/core on its own, this is legitimate. But it should be safe to make this change because the classes are always available if you have all of the required dependencies.

**Comment by srowen** [comment]
Am I crazy, or is it other way around? I don't see a catalyst dependency from core, and wouldn't actually expect that? But yeah I'm agreeing that this is an OK 'breaking' change to exclude.

**Comment by mccheah** [comment]
`sql/catalyst` depends on `sql/core`: https://github.com/apache/spark/blob/master/sql/core/pom.xml#L63

Though both are dependent on `spark-core`.

**Comment by srowen** [comment]
Oh, I kept thinking and looking at `core` not `sql/core`. Right. Well, same conclusion.

**Comment by gatorsmile** [comment]
retest this please

**Review by gatorsmile — COMMENTED** [review]
LGTM 

Also cc @zsxwing

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Comment by dongjoon-hyun** [comment]
Finally, Nice! Thank you, guys.

**Comment by leobenkel** [comment]
Hello,

I am trying to migrate from Spark 2.x to 3.x and replacing those 
```
import org.apache.spark.sql.sources.v2.writer.streaming.StreamWriter
import org.apache.spark.sql.sources.v2.{DataSourceOptions, StreamWriteSupport}
```
had been challenging. My investigation conducted me here, I read the code change and added `sql-catalyst` to my dependencies but can't figure out where those files went. 

Anyone could point me in the right direction ? Thanks !

**Comment by cloud-fan** [comment]
DS v2 has evolved a lot from Spark 2 to 3. I'm afraid you may need to stay with Spark 2 or rewrite your v2 source entirely.

**Comment by leobenkel** [comment]
> DS v2 has evolved a lot from Spark 2 to 3. I'm afraid you may need to stay with Spark 2 or rewrite your v2 source entirely.

I dont mind rewriting it. Would you have a tutorial I can follow ?

**Comment by cloud-fan** [comment]
You can always find it in tests :P

https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala

**Comment by leobenkel** [comment]
> You can always find it in tests :P
> 
> https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala

Super, I will take a look. Thank you so much for your help ! Have a wonderful rest of your day.
